### PR TITLE
Avoid writes without any data in `Write::write_all_vectored`

### PR DESCRIPTION
**Pull request by tmiasko**
Previously, when non-empty sequence of empty IoSlices have been provided
to `Write::write_all_vectored`, the buffers would be written as is with
`Write::write_vectored` and subsequently the return value `Ok(0)` would
be misinterpreted as an error.

Avoid writes without any data by advancing the buffers first. This
matches the documented behaviour of `Write::write_all_vectored`
and is analogous to what happens in `Write::write_all`.

### CODE REVIEW

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -1461,12 +1460,13 @@ pub trait Write {
     /// ```
     #[unstable(feature = "write_all_vectored", issue = "70436")]
     fn write_all_vectored(&mut self, mut bufs: &mut [IoSlice<'_>]) -> Result<()> {
+        bufs = IoSlice::advance(bufs, 0);
         while !bufs.is_empty() {
             match self.write_vectored(bufs) {
                 Ok(0) => {
                     return Err(Error::new(ErrorKind::WriteZero, "failed to write whole buffer"));
                 }
-                Ok(n) => bufs = IoSlice::advance(mem::take(&mut bufs), n),
```
Comment (Thomasdezeeuw) [inline src/libstd/io/mod.rs:1469]: It makes me happy this is possible now :)

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Muirrum** [comment]
@Thomasdezeeuw @tmiasko This is a triage bump.

**Comment by KodrAus** [comment]
I’m not very familiar with this API, but @tmiasko do you expect this unconditional advance to have any perf impact for the more common case of non-empty buffers?

**Comment by tmiasko** [comment]
I would expect the impact to be negligible compared to any I/O involved, and changes necessary regardless of the impact.

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -1461,12 +1460,13 @@ pub trait Write {
     /// ```
     #[unstable(feature = "write_all_vectored", issue = "70436")]
     fn write_all_vectored(&mut self, mut bufs: &mut [IoSlice<'_>]) -> Result<()> {
+        bufs = IoSlice::advance(bufs, 0);
```
Comment (KodrAus) [inline src/libstd/io/mod.rs:1463]: Let’s add a comment here explaining why this is needed.

Comment (tmiasko) [reply src/libstd/io/mod.rs:1439]: Added a comment explaining the motivation behind the initial advance.

**Pull request merged**